topology/probes/sriov/sriov.go

```diff
@@ -0,0 +1,177 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package sriov
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/skydive-project/skydive/common/fs"
+	"github.com/skydive-project/skydive/filters"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/graph"
+)
+
+// Probe is a Sriov probe designed to handle all vfs on server
+type Probe struct {
+	graph   *graph.Graph
+	root    *graph.Node
+	cancel  context.CancelFunc // cancel function
+	watcher *fs.SysProcFsWatcher
+}
+
+// Manager is a const to be used as a string to describe who handles sriov nodes
+const Manager = "vfio"
+
+// GenerateNameForNodeFromPhysInterfaceNameAndVFID - generates a name for vf based on physical device name and id of of this vf
+func GenerateNameForNodeFromPhysInterfaceNameAndVFID(physName string, id string) string {
+	return fmt.Sprintf("%s-vf-%s", physName, id)
+}
+
+func allZero(s []byte) bool {
+	for _, v := range s {
+		if v != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+func (h *Probe) getAllSriovNodes() []*graph.Node {
+	filter := filters.NewAndFilter(
+		filters.NewTermStringFilter("Manager", Manager),
+	)
+	m := graph.NewElementFilter(filter)
+	return h.graph.GetNodes(m)
+}
+
+func (h *Probe) periodicChecker(ctx context.Context) {
+	ticker := time.NewTicker(5 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			linkList, err := netlink.LinkList()
+			if err != nil {
+				logging.GetLogger().Errorf("unable to get link list : %s", err)
+			}
+			h.graph.Lock()
+			var addedNodes = make(map[string]bool)
+			for _, link := range linkList {
+				attrs := link.Attrs()
+				parentNode := h.graph.LookupFirstChild(h.root, graph.Metadata{"Name": attrs.Name, "IfIndex": int64(attrs.Index)})
+				if parentNode == nil {
+					continue
+				}
+				for _, vf := range attrs.Vfs {
+					if allZero(vf.Mac) {
+						continue
+					}
+					name := GenerateNameForNodeFromPhysInterfaceNameAndVFID(attrs.Name, strconv.Itoa(vf.ID))
+					logging.GetLogger().Debugf("try to process sriov nodes and vf : %s. MAC: %s, %b", attrs.Name, vf.Mac, vf.Mac)
+					h.addVfNode(name, parentNode, &vf)
+					addedNodes[name] = true
+				}
+				if len(attrs.Vfs) > 0 {
+					h.watcher.AddPath(fmt.Sprintf("/sys/class/net/%s/operstate", attrs.Name))
+				}
+			}
+			for _, node := range h.getAllSriovNodes() {
+				nodeName, _ := node.GetFieldString("Name")
+				if _, found := addedNodes[nodeName]; found {
+					continue
+				}
+				h.graph.DelNode(node)
+			}
+			h.graph.Unlock()
+		}
+	}
+}
+
+func (h *Probe) addVfNode(name string, parentNode *graph.Node, vf *netlink.VfInfo) {
```

Comment:
I think it would be better to add Metadata on the node created/seen by netlink instead of creating a new Node and I would create a link between the physical interface and the vfs one. I would add extra srio-v information in a SRIO-V section in the netlink vfs node metadata. Otherwise there is no way understand the relation between the vfs and the netlink interface. I could be also nice to add the PCI ID.

Reply:
> I think it would be better to add Metadata on the node created/seen by netlink instead of creating a new Node and I would create a link between the physical interface and the vfs one. I would add extra srio-v information in a SRIO-V section in the netlink vfs node metadata. Otherwise there is no way understand the relation between the vfs and the netlink interface. I could be also nice to add the PCI ID.

I believe there's some misunderstanding, here we don't create a new node which provides vf stuff. We use a node created/seen by netlink probe and simply adding a vf node as child of specific physical device. Would you like to use layer2 connection instead of parent connection ?
for example, if there's a physical device which provides a network interface ens20 and it has vf functionality then we add a node with name: ens2f0-vf-VFID.
About PCI ID, I'll do that.
About updating metadata for node seen/created by netlink, I think it would be excessive information which we may get analyzing graphs.

Reply:
Let's say you have a SRIO-V card, device `enp5s0f0`. It will be seen by the netlink probe. If you create two VFs on this device. Will will end up with 2 interfaces : `enp5s16` and `enp5s16f2`. Those will also be seen by the `netlink` probe.

The probe also creates 2 VF child nodes. These are a bit of a duplicate of the nodes created by `netlink`. Instead, we suggest to not create the VF node and use the 2 nodes already created by `netlink`, and link them to the physical device `enp5s0f0`.

Reply:
> Let's say you have a SRIO-V card, device `enp5s0f0`. It will be seen by the netlink probe. If you create two VFs on this device. Will will end up with 2 interfaces : `enp5s16` and `enp5s16f2`. Those will also be seen by the `netlink` probe.
> 
> The probe also creates 2 VF child nodes. These are a bit of a duplicate of the nodes created by `netlink`. Instead, we suggest to not create the VF node and use the 2 nodes already created by `netlink`, and link them to the physical device `enp5s0f0`.

Oh, I was stupid. It seems like netlink probe handles a lot of things I added here.......
I'll check and let you know...... I am sorry :(

Reply:
It seems there is only 3 attributes for this VF nodes : `Name`, `Manager` and `Type`. If we don't create new nodes and used the netlink ones, I don't see any attribute to add to them. The real value is on the edges from the physical devices : we can use a dedicated `RelationType`, such as `sriov`.

Reply:
> It seems there is only 3 attributes for this VF nodes : `Name`, `Manager` and `Type`. If we don't create new nodes and used the netlink ones, I don't see any attribute to add to them. The real value is on the edges from the physical devices : we can use a dedicated `RelationType`, such as `sriov`.

yes, let me revisit it again cause you broke my mind :) I realized I added some duplication :)

Reply:
It's a shame on me :(

Reply:
@sergeyglazyrindev-samsung  :-) I did not mean to. Thanks for working on it. It worked nicely on my test machine by the way.